modules/multiauth/lib/Controller/DiscoController.php

```diff
@@ -0,0 +1,137 @@
+use SimpleSAML\Logger;
+use SimpleSAML\Module;
+use SimpleSAML\Module\multiauth\Auth\Source\MultiAuth;
+use SimpleSAML\Session;
+use SimpleSAML\Utils;
+use SimpleSAML\XHTML\Template;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Controller class for the multiauth module.
+ *
+ * This class serves the different views available in the module.
+ *
+ * @package SimpleSAML\Module\multiauth
+ */
+class DiscoController
+{
+    /** @var \SimpleSAML\Configuration */
+    protected $config;
+
+    /** @var \SimpleSAML\Session */
+    protected $session;
+
+
+    /**
+     * Controller constructor.
+     *
+     * It initializes the global configuration and auth source configuration for the controllers implemented here.
+     *
+     * @param \SimpleSAML\Configuration              $config The configuration to use by the controllers.
+     * @param \SimpleSAML\Session                    $session The session to use by the controllers.
+     *
+     * @throws \Exception
+     */
+    public function __construct(
+        Configuration $config,
+        Session $session
+    ) {
+        $this->config = $config;
+        $this->session = $session;
+    }
+
+
+    /**
+     * This controller shows a list of authentication sources. When the user selects
+     * one of them if pass this information to the
+     * \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth class and call the
+     * delegateAuthentication method on it.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return \SimpleSAML\XHTML\Template|\Symfony\Component\HttpFoundation\RedirectResponse
+     *   An HTML template or a redirection if we are not authenticated.
+     */
+    public function discovery(Request $request)
+    {
+        // Retrieve the authentication state
+        $authStateId = $request->get('AuthState', null);
+        if (is_null($authStateId)) {
+            throw new Error\BadRequest('Missing AuthState parameter.');
+        }
+
+        /** @var array $state */
+        $state = Auth\State::loadState($authStateId, MultiAuth::STAGEID);
+
+        $as = null;
+        if (array_key_exists("\SimpleSAML\Auth\Source.id", $state)) {
+            $authId = $state["\SimpleSAML\Auth\Source.id"];
+
+            /** @var \SimpleSAML\Module\multiauth\Auth\Source\MultiAuth $as */
+            $as = Auth\Source::getById($authId);
+        }
+
+        $source = $request->get('source', null);
+        if (is_null($source)) {
+            foreach ($request->query->all() as $k => $v) {
+                $k = explode('-', $k, 2);
+                if (count($k) === 2 && $k[0] === 'src') {
+                    $source = base64_decode($k[1]);
+                }
+            }
+        }
+
+        if ($source !== null) {
+            if ($as !== null) {
+                $as->setPreviousSource($source);
+            }
+            MultiAuth::delegateAuthentication($source, $state);
```

Comment:
@jaimeperez would the StreamedResponse case apply here as well for the 'delegateAuthentication' call, like we did with the ADFS-module?

Reply:
Probably a `RunnableResponse` instead, in the `delegateAuthentication()` method itself. We need to call `authenticate()` on the auth source, and that's out of our control, so `RunnableResponse` would be the only alternative, I think. The entire block should then moved to a different method that we can pass to the response, since we need to consider the case where `authenticate()` actually returns, and call `Auth\Source::completeAuth($state);`.

The alternative would be to return a `RunnableResponse` passing the entire `delegateAuthentication()` method. That would be simpler, but less useful because then we can't really test how that method works.

Reply:
@jaimeperez I'm not sure I can follow what you're saying here, so I went with the quick-and-dirty solution.. Care to help me out and deploy your even-better solution?